### PR TITLE
:sparkles: : Create Discord API

### PR DESCRIPTION
**Pull request by utopi160**

### CODE REVIEW

**Comment by MoskalykA** [comment]
I should delete the webhook part of my pr and we would be complementary?

**Comment by MoskalykA** [comment]
https://github.com/NextCitizens/ncs_core/pull/58

**Comment by PABLO-1610** [comment]
> #58

The interesting thing is that his part is more focused on server-side (interaction with a Discord Bot etc) but your is more focused on the client one (Rich presence (...)), i think i will add your functions to this PR.

**Comment by MoskalykA** [comment]
> > #58
> 
> The interesting thing is that his part is more focused on server-side (interaction with a Discord Bot etc) but your is more focused on the client one (Rich presence (...)), i think i will add your functions to this PR.

I'll just delete the webhook part in this case

**Comment by utopi160** [comment]
Yess !
I'm sorry, I hadn't seen your pull request ....

**Comment by MoskalykA** [comment]
I have deleted my webhooks part

**Comment by MoskalykA** [comment]
> Yess ! I'm sorry, I hadn't seen your pull request ....

No, that's good. There is a full api

**Comment by PABLO-1610** [comment]
Hm, seems pretty good but, i think it could be better if we could have many instance of bot, in case of you need at least 2 bots doing 2 satisfying two differents things.

I recommand creating an object of DiscordBot, which will contains all functions related to it (eg: send message, retrieve users in guild etc.), the only parameter which will be passed through the constructor should be the bot token (it's really more secured than having it defined by "`API_Discord.token = "..."` .

Tell me what you guys think about this @MoskalykA @utopi160

**Comment by utopi160** [comment]
Mummm i see but how to organize the exports ?

> Hm, seems pretty good but, i think it could be better if we could have many instance of bot, in case of you need at least 2 bots doing 2 satisfying two differents things.
> 
> I recommand creating an object of DiscordBot, which will contains all functions related to it (eg: send message, retrieve users in guild etc.), the only parameter which will be passed through the constructor should be the bot token (it's really more secured than having it defined by "`API_Discord.token = "..."` .
> 
> Tell me what you guys think about this @MoskalykA @utopi160

**Comment by MoskalykA** [comment]
> Mummm i see but how to organize the exports ?

Yes object-oriented code

**Comment by MoskalykA** [comment]
> Mummm i see but how to organize the exports ?
> 
> > Hm, seems pretty good but, i think it could be better if we could have many instance of bot, in case of you need at least 2 bots doing 2 satisfying two differents things.
> > I recommand creating an object of DiscordBot, which will contains all functions related to it (eg: send message, retrieve users in guild etc.), the only parameter which will be passed through the constructor should be the bot token (it's really more secured than having it defined by "`API_Discord.token = "..."` .
> > Tell me what you guys think about this @MoskalykA @utopi160

I don't really know what you want?

**Comment by PABLO-1610** [comment]
@utopi160 take a look at : https://github.com/NextCitizens/ncs_core/tree/main/src/modules/players/server

You should create a new module and **then** a simple api which can interact with the module.

**Comment by utopi160** [comment]
> You should create a new module and **then** a simple api which can interact with the module.

Ok thank you !

**Comment by utopi160** [comment]
I don't really see how the API is organized ?
image

**Comment by MoskalykA** [comment]
> I don't really see how the API is organized ? image

You should make an object-oriented structure in this style:
ob

**Comment by utopi160** [comment]
But it's not okay ?
https://github.com/utopi160/ncs_core/tree/main/src/class/Bot/server
https://github.com/utopi160/ncs_core/tree/main/src/modules/bot

**Comment by MoskalykA** [comment]
> But it's not okay ? https://github.com/utopi160/ncs_core/tree/main/src/class/Bot/server https://github.com/utopi160/ncs_core/tree/main/src/modules/bot

That's good, you could also make a file by function

**Comment by utopi160** [comment]
But I don't see how to relate to API_DiscordBot, an idea ?

**Comment by MoskalykA** [comment]
> But I don't see how to relate to API_DiscordBot, an idea ?

image
I would have done a bit like that but I would have put:
    Bot = NCSBot
    Bots = MOD_Bots

**Comment by utopi160** [comment]
Ok thank you !!

> > But I don't see how to relate to API_DiscordBot, an idea ?
> 
> image I would have done a bit like that but I would have put: Bot = NCSBot Bots = MOD_Bots

File: src/class/Bot/server/NCSBot.lua [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+                return ("https://cdn.discordapp.com/avatars/%s/%s.gif"):format(discordId, player.avatar)
+            else
+                return ("https://cdn.discordapp.com/avatars/%s/%s.png"):format(discordId, player.avatar)
+            end
+        else
+            return ("https://cdn.discordapp.com/embed/avatars/0.png")
+        end
+    else
+        return ("https://cdn.discordapp.com/embed/avatars/0.png")
+    end
```
Comment (MoskalykA) [inline src/class/Bot/server/NCSBot.lua:107]:
```suggestion
    if (player and player.avatar) then
        if (player.avatar:sub(2, 2) == "_") then -- Determine if the avatar is PNG or GIF
            return ("https://cdn.discordapp.com/avatars/%s/%s.gif"):format(discordId, player.avatar)
        else
            return ("https://cdn.discordapp.com/avatars/%s/%s.png"):format(discordId, player.avatar)
        end
    else
        return ("https://cdn.discordapp.com/embed/avatars/0.png")
    end
```

File: src/class/Bot/server/NCSBot.lua [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+---@param discordId string
+---@public
+---@return table
+function NCSBot:get_player(discordId)
+    return (json.decode(self:performDiscord(_NCSEnum.MethodType.GET, ("users/%s"):format(discordId))))
+end
+
+---get_guild
+---@public
+---@return table
```
Comment (MoskalykA) [inline src/class/Bot/server/NCSBot.lua:55]:
```suggestion
---get_guild
---@param guildId string
---@public
---@return table
```

File: src/class/Bot/server/NCSBot.lua [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+function NCSBot:get_guild_roles(guildId)
+    local guild = self:get_guild(guildId)
+
+    return (guild and guild.roles or {})
+end
+
+---get_guild_roles_count
+---@param guildId string
+---@public
+---@return table
```
Comment (MoskalykA) [inline src/class/Bot/server/NCSBot.lua:155]:
```suggestion
---@return number
```

File: src/class/Bot/server/NCSBot.lua [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+function NCSBot:get_guild_roles_count(guildId)
+    local guild = self:get_guild(guildId)
+
+    return (guild and guild.roles and #guild.roles or 0)
+end
+
+---get_guild_members_count
+---@param guildId string
+---@public
+---@return table
```
Comment (MoskalykA) [inline src/class/Bot/server/NCSBot.lua:165]:
```suggestion
---@return number
```

File: src/class/Bot/server/NCSBot.lua [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+function NCSBot:get_guild_members_count(guildId)
+    local guild = self:get_guild_count(guildId)
+
+    return (guild and guild.approximate_member_count or 0)
+end
+
+---get_guild_online_members_count
+---@param guildId string
+---@public
+---@return table
```
Comment (MoskalykA) [inline src/class/Bot/server/NCSBot.lua:175]:
```suggestion
---@return number
```

File: src/class/Bot/server/NCSBot.lua [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                return ("https://cdn.discordapp.com/icons/%s/%s.gif"):format(guildId, guild.icon)
+            else
+                return ("https://cdn.discordapp.com/icons/%s/%s.png"):format(guildId, guild.icon)
+            end
+        else
+            return ("https://cdn.discordapp.com/embed/avatars/0.png")
+        end
+    else
+        return ("https://cdn.discordapp.com/embed/avatars/0.png")
+    end
```
Comment (MoskalykA) [inline src/class/Bot/server/NCSBot.lua:139]:
```suggestion
    if (guild and guild.icon) then
        if (guild.icon:sub(2, 2) == "_") then -- Determine if the icon is PNG or GIF
            return ("https://cdn.discordapp.com/icons/%s/%s.gif"):format(guildId, guild.icon)
        else
            return ("https://cdn.discordapp.com/icons/%s/%s.png"):format(guildId, guild.icon)
        end
    else
        return ("https://cdn.discordapp.com/embed/avatars/0.png")
    end
```

**Review by MoskalykA — CHANGES_REQUESTED** [review]

**Comment by utopi160** [comment]
It's not OK ?

**Review by PABLO-1610 — APPROVED** [review]
Awesome PR 🤩 Thanks a lot

**Pull request merged**